gui/RespiraConnectedDevice.cpp

```diff
@@ -0,0 +1,83 @@
+#include "../common/third_party/nanopb/pb_common.h"
+#include "../common/third_party/nanopb/pb_decode.h"
+#include "../common/third_party/nanopb/pb_encode.h"
+#include "connected_device.h"
+#include <QSerialPort>
+
+class RespiraConnectedDevice : public ConnectedDevice {
+
+  RespiraConnectedDevice(QString portName, uint32_t write_timeout_ms,
+                         uint32_t interframe_timeout_ms)
+      : m_serial(new QSerialPort()), m_writeTimeout_ms(write_timeout_ms),
+        m_interframeTimeout_ms(interframe_timeout_ms) {
+    m_serial->setPortName(portName);
+    m_serial->setBaudRate(QSerialPort::Baud115200);
+    m_serial->setDataBits(QSerialPort::Data8);
+    m_serial->setParity(QSerialPort::NoParity);
+    m_serial->setStopBits(QSerialPort::OneStop);
+    m_serial->setFlowControl(QSerialPort::NoFlowControl);
+    m_serial->open(QIODevice::ReadWrite);
+  }
+
+public:
+  void SendGuiStatus(const GuiStatus &gui_status) override {
+
+    uint8_t tx_buffer[GuiStatus_size];
+    pb_ostream_t stream = pb_ostream_from_buffer(tx_buffer, sizeof(tx_buffer));
+    if (!pb_encode(&stream, GuiStatus_fields, &gui_status)) {
+      // TODO: Serialization failure; log an error or raise an alert.
+      return;
+    }
+
+    m_serial->write((const char *)tx_buffer, stream.bytes_written);
+
+    if (!m_serial->waitForBytesWritten(m_writeTimeout_ms)) {
+      // TODO communication failure, port closed? Log an error and raise an
+      // alert
+      return;
+    }
+  }
+
+  void ReceiveControllerStatus(ControllerStatus *controller_status) override {
+    // Drop everything we have in buffer right now as we are not sure if this is
+    // a full frame or a part of it.
+    // We drop everything that comes in until we get at least half of interframe
+    // silence
+    if (m_serial->bytesAvailable() > 0) {
+      while (m_serial->waitForReadyRead(m_interframeTimeout_ms / 2)) {
+        m_serial->clear(QSerialPort::Input);
+      }
+    }
+
+    if (m_serial->waitForReadyRead(m_interframeTimeout_ms)) {
+      QByteArray responseData = m_serial->readAll();
+
+      // continue reading characters until we read all we expect or we don't see
+      // any data for long enough time to estimate failure
+      while (responseData.length() < ControllerStatus_size &&
+             m_serial->waitForReadyRead(m_interframeTimeout_ms / 2)) {
+        responseData += m_serial->readAll();
+      }
+
+      if (ControllerStatus_size == responseData.length()) {
+        pb_istream_t stream = pb_istream_from_buffer(
+            (const uint8_t *)responseData.data(), responseData.length());
+        if (!pb_decode(&stream, ControllerStatus_fields, controller_status)) {
+          // TODO: Log an error.
+        }
+      } else {
+        // TODO we have read some garbage here. Log an error. Raise an alert
+        return;
```

Comment:
nit: instead of a bunch of conditions "if (blah) { foo } else {return}", I'd suggest to invert and get: if (!blah) return; foo - this way foo has reduced nesting which is easier to read.